PrepareFiles.py

```diff
@@ -1,106 +1,164 @@
+if "DIGITAL_METADATA_API" in os.environ:
     s3_bucket_get = os.environ['S3_BUCKET_NAME_GET']
-else:
-    s3_bucket_get = PreparedFiles_config.S3_BUCKET_NAME_GET
-if "S3_BUCKET_NAME_PUT" in os.environ:
     s3_bucket_put = os.environ['S3_BUCKET_NAME_PUT']
+    metadata_api = os.environ['DIGITAL_METADATA_API']
+    role_arn = os.environ['S3_CROSS_ACCOUNT_ROLE_ARN']
+    client = boto3.client('sts')
+    sts_response = client.assume_role(
+        RoleArn=role_arn,
+        RoleSessionName='cross_acct_lambda',
+        DurationSeconds=900
+    )
+    ACCESS_KEY = sts_response['Credentials']['AccessKeyId']
+    SECRET_KEY = sts_response['Credentials']['SecretAccessKey']
+    SESSION_TOKEN = sts_response['Credentials']['SessionToken']
+    s3_client = boto3.client(
+        's3',
+        aws_access_key_id=ACCESS_KEY,
+        aws_secret_access_key=SECRET_KEY,
+        aws_session_token=SESSION_TOKEN
+    )
 else:
-    s3_bucket_put = PreparedFiles_config.S3_BUCKET_NAME_PUT
-if "Digital_MetaData_API" in os.environ:
-    metadata_api = os.environ['Digital_MetaData_API']
-else:
-    metadata_api = PreparedFiles_config.Digital_MetaData_API
+    s3_bucket_get = 0
+    s3_bucket_put = 0
+    metadata_api = 0
+    role_arn = 0
 
 
 def get_replica(rid):
-    with urllib.request.urlopen(metadata_api+rid) as content_file:
+    with urllib.request.urlopen(metadata_api + 'replicas/' + rid) as content_file:
         json_content = content_file.read()
     content = json.loads(json_content)
     replica = Replica(content)
     return replica
 
 
 class Replica:
-    def __init__( self, content ):
+    def __init__(self, content):
         self.replica_data = content
 
-    def process_files(self, delivery_type, max_deliveryfile_size, reference):
-        # TODO Code the ZIP case. Can any of the PDF case be generalised?
+    def process_files(self, delivery_type, max_deliveryfile_size, reference, iaid):
         status = False
         if delivery_type == "pdf":
-            status = self._create_pdf(max_deliveryfile_size, reference)
-        # else:
-        #     create_zip(cal_avg_size)
+            status = self._create_pdf(max_deliveryfile_size, reference, iaid)
+        else:
+            status = self._create_zip(max_deliveryfile_size, reference, iaid)
         return status
 
-    def _create_image_list(self, max_deliveryfile_size, data):
+    def _create_file_list(self, max_deliveryfile_size, data):
         output_file_list = []
         size = 0
-        this_pdf = []
+        this_file = []
         length = len(data)
         count = 1
         for file in data:
-            this_pdf.append(file['name'])
-            size = file['size'] + size
-            if size > max_deliveryfile_size:
-                output_file_list.append(this_pdf)
+            if length == count:
+                this_file.append(file['name'])
+                output_file_list.append(this_file)
+            elif (file['size'] + data[count]['size']) > max_deliveryfile_size:
```

Comment:
What are we testing here? I assume the logic should be "If the file we are about to add means that the maximum file size will be exceeded, add it, but then start a new PDF", however we're getting the size of the file we're about to add (`file['size']`) adding the size of the next JPG in the list (`data[count]['size']`) and checking that it doesn't exceed max_deliveryfile_size?
Is there an error here?

Reply:
No error. I had an instance where one file was 49mb and the next file was 130mb. The previous logic would loop through once, then would say, that's fine it's under 50mb, then loop through the second time, add the next file of 130mb, and then would say, we are now over 50mb so finish this batch. But that means that batch would be 179mb!